Index out of bounds while checking function arguments

### Code

```Rust
fn f() {
    g((), ());
}

pub fn g(a1: (), a2: bool, a3: bool, a4: bool, a5: bool, a6: ()) {}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2bcda868a0f4634e13b9d8b77a7f5832)

### Meta
`rustc --version --verbose`:
```
rustc 1.63.0-nightly (cd282d7f7 2022-05-18)
binary: rustc
commit-hash: cd282d7f75da9080fda0f1740a729516e7fbec68
commit-date: 2022-05-18
host: x86_64-unknown-linux-gnu
release: 1.63.0-nightly
LLVM version: 14.0.4
```

### Error output

```
thread 'rustc' panicked at 'index out of bounds: the len is 2 but the index is 4', compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:818:59

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: rustc 1.63.0-nightly (cd282d7f7 2022-05-18) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `f`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `rustc_bug`
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0:     0x7fc08d000fcd - std::backtrace_rs::backtrace::libunwind::trace::h792c30771d1dce4e
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7fc08d000fcd - std::backtrace_rs::backtrace::trace_unsynchronized::h20bd96032223b984
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fc08d000fcd - std::sys_common::backtrace::_print_fmt::h55ce8a26aee332a3
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7fc08d000fcd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1e608165fd05792c
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7fc08d05cc5c - core::fmt::write::hdc9f366f1ec7f5cc
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/fmt/mod.rs:1196:17
   5:     0x7fc08cff2741 - std::io::Write::write_fmt::hae36452b87a7144a
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/io/mod.rs:1654:15
   6:     0x7fc08d003ce5 - std::sys_common::backtrace::_print::hbfb86fe090e52690
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7fc08d003ce5 - std::sys_common::backtrace::print::h4abb1fed1562e0f8
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7fc08d003ce5 - std::panicking::default_hook::{{closure}}::hed4b61614b3d60d1
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:295:22
   9:     0x7fc08d003959 - std::panicking::default_hook::h787a125a3d53fabc
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:314:9
  10:     0x7fc08d7c9b41 - rustc_driver[af3c34e99bc7132d]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7fc08d0044b6 - std::panicking::rust_panic_with_hook::he936e1666803a13d
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:702:17
  12:     0x7fc08d0042b7 - std::panicking::begin_panic_handler::{{closure}}::h5330690fb8e8d9da
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:588:13
  13:     0x7fc08d001484 - std::sys_common::backtrace::__rust_end_short_backtrace::he03416277ca16606
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7fc08d003fe9 - rust_begin_unwind
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:584:5
  15:     0x7fc08cfc9263 - core::panicking::panic_fmt::hcb1f1aab333b06ef
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/panicking.rs:142:14
  16:     0x7fc08cfc91a2 - core::panicking::panic_bounds_check::h2a4a0e6246b57d65
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/panicking.rs:84:5
  17:     0x7fc08ecafdc6 - <rustc_typeck[a51bc4bda11be944]::check::fn_ctxt::FnCtxt>::check_argument_types
  18:     0x7fc08ec7833c - <rustc_typeck[a51bc4bda11be944]::check::fn_ctxt::FnCtxt>::check_call
  19:     0x7fc08ecc6399 - <rustc_typeck[a51bc4bda11be944]::check::fn_ctxt::FnCtxt>::check_expr_kind
  20:     0x7fc08ecc5210 - <rustc_typeck[a51bc4bda11be944]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  21:     0x7fc08ecb4099 - <rustc_typeck[a51bc4bda11be944]::check::fn_ctxt::FnCtxt>::check_stmt
  22:     0x7fc08ecb571d - <rustc_typeck[a51bc4bda11be944]::check::fn_ctxt::FnCtxt>::check_block_with_expected
  23:     0x7fc08ecc64aa - <rustc_typeck[a51bc4bda11be944]::check::fn_ctxt::FnCtxt>::check_expr_kind
  24:     0x7fc08ecc5210 - <rustc_typeck[a51bc4bda11be944]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  25:     0x7fc08ede1640 - rustc_typeck[a51bc4bda11be944]::check::check::check_fn
  26:     0x7fc08ed46bbc - <rustc_infer[73d5c1fd1d2bb465]::infer::InferCtxtBuilder>::enter::<&rustc_middle[29414c9708d2f2c6]::ty::context::TypeckResults, <rustc_typeck[a51bc4bda11be944]::check::inherited::InheritedBuilder>::enter<rustc_typeck[a51bc4bda11be944]::check::typeck_with_fallback<rustc_typeck[a51bc4bda11be944]::check::typeck::{closure#0}>::{closure#1}, &rustc_middle[29414c9708d2f2c6]::ty::context::TypeckResults>::{closure#0}>
  27:     0x7fc08ed03c90 - rustc_typeck[a51bc4bda11be944]::check::typeck
  28:     0x7fc08fd28a70 - <rustc_query_system[c5b167391a1fc8c0]::dep_graph::graph::DepGraph<rustc_middle[29414c9708d2f2c6]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[29414c9708d2f2c6]::ty::context::TyCtxt, rustc_span[59d1c307ce41abbc]::def_id::LocalDefId, &rustc_middle[29414c9708d2f2c6]::ty::context::TypeckResults>
  29:     0x7fc08f2baee1 - rustc_query_system[c5b167391a1fc8c0]::query::plumbing::try_execute_query::<rustc_query_impl[d1fe922fcf7cfd91]::plumbing::QueryCtxt, rustc_query_system[c5b167391a1fc8c0]::query::caches::DefaultCache<rustc_span[59d1c307ce41abbc]::def_id::LocalDefId, &rustc_middle[29414c9708d2f2c6]::ty::context::TypeckResults>>
  30:     0x7fc08f28b188 - <rustc_query_impl[d1fe922fcf7cfd91]::Queries as rustc_middle[29414c9708d2f2c6]::ty::query::QueryEngine>::typeck
  31:     0x7fc08eda37d8 - <rustc_middle[29414c9708d2f2c6]::hir::map::Map>::par_body_owners::<rustc_typeck[a51bc4bda11be944]::check::typeck_item_bodies::{closure#0}>
  32:     0x7fc08fad650c - rustc_typeck[a51bc4bda11be944]::check::typeck_item_bodies
  33:     0x7fc08fd45ee9 - <rustc_query_system[c5b167391a1fc8c0]::dep_graph::graph::DepGraph<rustc_middle[29414c9708d2f2c6]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[29414c9708d2f2c6]::ty::context::TyCtxt, (), ()>
  34:     0x7fc08fdec61a - rustc_query_system[c5b167391a1fc8c0]::query::plumbing::try_execute_query::<rustc_query_impl[d1fe922fcf7cfd91]::plumbing::QueryCtxt, rustc_query_system[c5b167391a1fc8c0]::query::caches::DefaultCache<(), ()>>
  35:     0x7fc08fe15441 - rustc_query_system[c5b167391a1fc8c0]::query::plumbing::get_query::<rustc_query_impl[d1fe922fcf7cfd91]::queries::typeck_item_bodies, rustc_query_impl[d1fe922fcf7cfd91]::plumbing::QueryCtxt>
  36:     0x7fc08fb0b0b3 - <rustc_session[8640123feec5f500]::session::Session>::time::<(), rustc_typeck[a51bc4bda11be944]::check_crate::{closure#7}>
  37:     0x7fc08faf5ddb - rustc_typeck[a51bc4bda11be944]::check_crate
  38:     0x7fc08f88fee7 - rustc_interface[3e05e90922109e97]::passes::analysis
  39:     0x7fc08fd420b5 - <rustc_query_system[c5b167391a1fc8c0]::dep_graph::graph::DepGraph<rustc_middle[29414c9708d2f2c6]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[29414c9708d2f2c6]::ty::context::TyCtxt, (), core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>
  40:     0x7fc08fde25b1 - rustc_query_system[c5b167391a1fc8c0]::query::plumbing::try_execute_query::<rustc_query_impl[d1fe922fcf7cfd91]::plumbing::QueryCtxt, rustc_query_system[c5b167391a1fc8c0]::query::caches::DefaultCache<(), core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>>
  41:     0x7fc08fe2791e - rustc_query_system[c5b167391a1fc8c0]::query::plumbing::get_query::<rustc_query_impl[d1fe922fcf7cfd91]::queries::analysis, rustc_query_impl[d1fe922fcf7cfd91]::plumbing::QueryCtxt>
  42:     0x7fc08f870bf7 - <rustc_interface[3e05e90922109e97]::passes::QueryContext>::enter::<rustc_driver[af3c34e99bc7132d]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>
  43:     0x7fc08f85b9c8 - <rustc_interface[3e05e90922109e97]::interface::Compiler>::enter::<rustc_driver[af3c34e99bc7132d]::run_compiler::{closure#1}::{closure#2}, core[d4fcbe885e6a87a3]::result::Result<core[d4fcbe885e6a87a3]::option::Option<rustc_interface[3e05e90922109e97]::queries::Linker>, rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>
  44:     0x7fc08f88433f - rustc_span[59d1c307ce41abbc]::with_source_map::<core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>, rustc_interface[3e05e90922109e97]::interface::create_compiler_and_run<core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>, rustc_driver[af3c34e99bc7132d]::run_compiler::{closure#1}>::{closure#1}>
  45:     0x7fc08f85c734 - rustc_interface[3e05e90922109e97]::interface::create_compiler_and_run::<core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>, rustc_driver[af3c34e99bc7132d]::run_compiler::{closure#1}>
  46:     0x7fc08f859ec2 - <scoped_tls[cbc0d094596bada1]::ScopedKey<rustc_span[59d1c307ce41abbc]::SessionGlobals>>::set::<rustc_interface[3e05e90922109e97]::interface::run_compiler<core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>, rustc_driver[af3c34e99bc7132d]::run_compiler::{closure#1}>::{closure#0}, core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>
  47:     0x7fc08f87127f - std[5ea0d33796e47bea]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3e05e90922109e97]::util::run_in_thread_pool_with_globals<rustc_interface[3e05e90922109e97]::interface::run_compiler<core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>, rustc_driver[af3c34e99bc7132d]::run_compiler::{closure#1}>::{closure#0}, core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>::{closure#0}, core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>
  48:     0x7fc08f8713b9 - <<std[5ea0d33796e47bea]::thread::Builder>::spawn_unchecked_<rustc_interface[3e05e90922109e97]::util::run_in_thread_pool_with_globals<rustc_interface[3e05e90922109e97]::interface::run_compiler<core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>, rustc_driver[af3c34e99bc7132d]::run_compiler::{closure#1}>::{closure#0}, core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>::{closure#0}, core[d4fcbe885e6a87a3]::result::Result<(), rustc_errors[8ec3179ad5714546]::ErrorGuaranteed>>::{closure#1} as core[d4fcbe885e6a87a3]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  49:     0x7fc08d00e3d3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h80876dac74f8f7f5
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/alloc/src/boxed.rs:1872:9
  50:     0x7fc08d00e3d3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h9cd29f95cc73ffbe
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/alloc/src/boxed.rs:1872:9
  51:     0x7fc08d00e3d3 - std::sys::unix::thread::Thread::new::thread_start::hc9488b0bdef72094
                               at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/sys/unix/thread.rs:108:17
  52:     0x7fc08cf23609 - start_thread
                               at /build/glibc-eX1tMB/glibc-2.31/nptl/pthread_create.c:477:8
  53:     0x7fc08ce3d293 - clone
  54:                0x0 - <unknown>
```

</p>
</details>



<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"compiler-errors"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

Also ICEs on beta. Output on stable:

```
error[E0061]: this function takes 6 arguments but 2 arguments were supplied
  --> src/main.rs:2:5
   |
2  |     g((), ());
   |     ^ --  -- supplied 2 arguments
   |     |
   |     expected 6 arguments
   |
note: function defined here
  --> src/main.rs:5:8
   |
5  |   pub fn g(
   |          ^
6  |       a1: (),
   |       ------
7  |       a2: bool,
   |       --------
8  |       a3: bool,
   |       --------
9  |       a4: bool,
   |       --------
10 |       a5: bool,
   |       --------
11 | /     a6: ()
12 | | ) -> () {
   | |_
```

@rustbot label regression-from-stable-to-beta

--- a/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
+++ b/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
@@ -768,7 +768,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 let second_input_ty =
                                     self.resolve_vars_if_possible(expected_input_tys[second_idx]);
                                 let third_input_ty =
-                                    self.resolve_vars_if_possible(expected_input_tys[second_idx]);
+                                    self.resolve_vars_if_possible(expected_input_tys[third_idx]);
                                 let span = if third_idx < provided_arg_count {
                                     let first_arg_span = provided_args[first_idx].span;
                                     let third_arg_span = provided_args[third_idx].span;
@@ -809,16 +809,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             }
                             missing_idxs => {
                                 let first_idx = *missing_idxs.first().unwrap();
-                                let second_idx = *missing_idxs.last().unwrap();
+                                let last_idx = *missing_idxs.last().unwrap();
                                 // NOTE: Because we might be re-arranging arguments, might have extra arguments, etc.
                                 // It's hard to *really* know where we should provide this error label, so this is a
                                 // decent heuristic
-                                let span = if first_idx < provided_arg_count {
+                                let span = if last_idx < provided_arg_count {
                                     let first_arg_span = provided_args[first_idx].span;
-                                    let second_arg_span = provided_args[second_idx].span;
+                                    let last_arg_span = provided_args[last_idx].span;
                                     Span::new(
                                         first_arg_span.lo(),
-                                        second_arg_span.hi(),
+                                        last_arg_span.hi(),
                                         first_arg_span.ctxt(),
                                         None,
                                     )

--- a/src/test/ui/argument-suggestions/issue-97197.rs
+++ b/src/test/ui/argument-suggestions/issue-97197.rs
@@ -1,0 +1,6 @@
+fn main() {
+    g((), ());
+    //~^ ERROR this function takes 6 arguments but 2 arguments were supplied
+}
+
+pub fn g(a1: (), a2: bool, a3: bool, a4: bool, a5: bool, a6: ()) -> () {}

--- a/src/test/ui/argument-suggestions/issue-97197.stderr
+++ b/src/test/ui/argument-suggestions/issue-97197.stderr
@@ -1,0 +1,19 @@
+error[E0061]: this function takes 6 arguments but 2 arguments were supplied
+  --> $DIR/issue-97197.rs:2:5
+   |
+LL |     g((), ());
+   |     ^-------- multiple arguments are missing
+   |
+note: function defined here
+  --> $DIR/issue-97197.rs:6:8
+   |
+LL | pub fn g(a1: (), a2: bool, a3: bool, a4: bool, a5: bool, a6: ()) -> () {}
+   |        ^ ------  --------  --------  --------  --------  ------
+help: provide the arguments
+   |
+LL |     g((), {bool}, {bool}, {bool}, {bool}, ());
+   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0061`.

--- a/src/test/ui/argument-suggestions/missing_arguments.stderr
+++ b/src/test/ui/argument-suggestions/missing_arguments.stderr
@@ -293,7 +293,7 @@ error[E0061]: this function takes 5 arguments but 2 arguments were supplied
   --> $DIR/missing_arguments.rs:39:3
    |
 LL |   complex(   1,                     ""   );
-   |   ^^^^^^^--------------------------------- three arguments of type `f32`, `i32`, and `i32` are missing
+   |   ^^^^^^^--------------------------------- three arguments of type `f32`, `i32`, and `f32` are missing
    |
 note: function defined here
   --> $DIR/missing_arguments.rs:7:4

